Update nuget.exe push to Azdo Artifact example (#12425)
MicrosoftDocs/azure-devops-docs: docs/artifacts/includes/nuget/nuget-publish-endpoint.md

    :::image type="content" source="../../media/nuget-connect-feed.png" alt-text="NuGet.exe feed connection":::

> 1. When using a local development environment, you must have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).

Reviewer: ```suggestion
1. When using a local development environment, it is recommended to have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).
```

Technically the PAT can be saved in a nuget.config as a package source credential, it's just better to use the credential provider.

    :::image type="content" source="../../media/nuget-connect-feed.png" alt-text="NuGet.exe feed connection":::

1. When using a local development environment, it is recommended to have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages [see examples](#examples).

    1. Download the [latest NuGet version](https://www.nuget.org/downloads).

    1. Follow the instructions in the **Project setup** to add a config file. 

        :::image type="content" source="../../media/project-setup.png" alt-text="Project setup":::

> 1. For Azure hosted builds that interact with Azure Artifacts feeds, you must use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task for the authentication [(see examples)](#examples).

Author: Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).

1. For Github actions builds that interact with Azure Artifacts feeds, you may use [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication [(see examples)](#examples).

> 1. For other hosted builds that interact with Azure Artifacts feeds, you may use [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication [(see examples)](#examples), but please exercise caution because it may leak your PAT as part of build environement log collection or source control checkout, artifact upload accidentally.

Author: Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).

## Examples

### Pushing from a local development environment

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Author: Sorry, my current PR doesn't reflect latest changes in private branch we built for preview. I should have been more explicit.

We made some adjustments in private branch so it may look different from my current branch
Could you look at private branch instead of changes in this PR?

**nuget.exe push example**: https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782
[Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) requirement added into [Prerequisites section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#prerequisites).
We added push from other [external sources (for teamcity, jenkins etc ..) section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#publish-packages-from-external-sources) at bottom.

**Push from pipeline example**: We added script example as alternative to nuget.command task in pipeline example: https://review.docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&branch=pr-en-us-4782&tabs=yaml#publish-a-package

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Reviewer 2: Can we review that PR instead?

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Reviewer 2: I don't have access to view the preview.

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Author: Private PR link is: https://github.com/MicrosoftDocs/azure-devops-docs-pr/pull/4782
If you get 404 error then you need to join `MicrosoftDocs` org which is pretty straight forward, [click step 3 here](https://review.docs.microsoft.com/en-us/help/get-started/setup-github?branch=main#3-join-microsoftdocs-and-other-organizations).  If you have any comment let's continue on private PR.

Actually, change is already live now https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops, so preview link is deleted. If you want to compare, then [previous version is here](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?branch=pazand-patch-1&view=azure-devops).

This PR show merged because private PR based on this, it was automatic step once private PR go live.
Let's continue discussion on private PR.

Thank you.

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Reviewer 2: That doesn't seem to mentioned the authenticate task at all, is it on purpose?

> When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.

Author: Authenticate task example is here [Azure Pipelines (YAML/Classic)](https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml#publish-a-package) since it's related only to Azire pipeline. It's how they organize their docs.

nuget push MyPackage.5.0.2.nupkg -src https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json AZ
```

### Pushing from Azure hosted build

Please note here the API key "AZ" is only used as a placeholder.

  ```Command
    - task: NuGetAuthenticate@1
      inputs:
        nuGetServiceConnections: MyServiceConnection_ExternalServer
        
    - script: |
        nuget push nupkgs/foo.1.1.5.nupkg -src https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json AZ
      displayName: "Push"          
  ```

### Pushing from GitHub Actions

Please note here the API key "AZ" is only used as a placeholder.

```Command
    - uses: actions/setup-dotnet@v2
      with:
        source-url: https://pkgs.dev.azure.com/<your-organization>/_packaging/<your-feed-name>/nuget/v3/index.json
      env:
        NUGET_AUTH_TOKEN: ${{secrets.AZURE_DEVOPS_PAT}} # Note, create a secret with this name in Settings
    
    - name: Push
      run: |
          nuget push nupkgs/mypackage.1.1.3.nupkg -src https://pkgs.dev.azure.com/<your-organization>/_packaging/<your-feed-name>/nuget/v3/index.json AZ
```

### Pushing from other hosted builds

> When pushing from any other hosted builds (than local host, Azure hosted build, Github actions) to Azure Artifacts, your pipeline can use the [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication, the API key and UserName "AZ" are only used as a placeholder. Use this with caution, it may leak your PAT as part of build environement log collection or source control checkout, artifact upload accidentally.

Author: Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).

    :::image type="content" source="../../media/nuget-connect-feed.png" alt-text="NuGet.exe feed connection":::

> 1. When using a local development environment, it is recommended to have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).

Author: Sure, fixed.

    1. Download the [latest NuGet version](https://www.nuget.org/downloads).

    1. Follow the instructions in the **Project setup** to add a config file. 

        :::image type="content" source="../../media/project-setup.png" alt-text="Project setup":::

1. For Azure hosted builds that interact with Azure Artifacts feeds, you must use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task for the authentication(see examples section).

1. For Github actions builds that interact with Azure Artifacts feeds, you may use [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication(see examples section).

> 1. For other hosted builds that interact with Azure Artifacts feeds, you may use [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication(see examples section), but please exercise caution because it may leak your PAT as part of build environement log collection or source control checkout, artifact upload accidentally.

Author: I added `other hosted builds` section because there're other hosting such as Teamcity, Jenkins etc which may don't support [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task, but still want to publish nupkg package to Azure Artifacts.

## Examples

> - In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.

Reviewer 2: Another idea is to put `ignored` as the API-key value. 

Makes it very explicit that the vlaue is irrelevant.

Author: Good input. Still `AZ` is short, also subtly express it's Azure only quirk. Let's keep it as it's for now, I'm welcome to other ideas.

Reviewer 2: I think you might have misunderstood me. 

If you put the value in the example, it'll be obvious what is going on and you wouldn't need the sentence explaining the placeholder.

Author: Removed the `(it can any value)`

> ```dotnetcli

Reviewer: I think `Command` is correct. At least, that's what they use in other docs pages: https://github.com/MicrosoftDocs/azure-devops-docs/blame/main/docs/artifacts/get-started-nuget.md#L83-L92

> nuget push MyPackage.5.0.2.nupkg -src https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json AZ

Reviewer 2: I think you're missing an api key option here.

Reviewer 2: I see. As long as it works I guess. :D

```

- In the example below, pushing "MyPackage" version "1.1.5" to an Azure Artifacts feed from an Azure hosted build, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires your pipeline to use the [NuGet Authenticate task](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate?view=azure-devops).

  ```cli
    - task: NuGetAuthenticate@1
      inputs:
        nuGetServiceConnections: MyServiceConnection_ExternalServer
        
    - script: |

>         dotnet build mypackage/mypackage.csproj -c Release

Reviewer 2: nit: I'd just include the push example. 

It's a slippery slope once you add all the steps.

Reviewer 2: Running `dotnet build`, `dotnet pack`  in the same script is not the best idea either as they are implicit. 

You could always just run `dotnet pack`, but I still think it isn't necessary. 

Curious what others think.

Reviewer: I agree. Many developers will want restore, build, pack, and push in different steps, so that if something fails, it's much more obvious what step it failed at, before even reading the logs.

Also, I find it odd that build and pack is done with the `dotnet` cli, but then push with nuget.exe.  If you're going to build with the dotnet CLI, you might as well also push with it.

Honestly, I don't think there should be separate pages for nuget.exe and dotnet cli. I think there should just be a single "push from cli" page, and it should only use the dotnet cli.  Let's not encourage customers to use nuget.exe.

Author: Sure, I removed unnecessary `dotnet build` from all examples.

Author: > Honestly, I don't think there should be separate pages for nuget.exe and dotnet cli. I think there should just be a single "push from cli" page, and it should only use the dotnet cli. Let's not encourage customers to use nuget.exe.

If I remember correctly there're some edge cases like mono which still use nuget.exe, so keep it separate for now. Removing a page easier thing to do next time, so let's deal next time.

        dotnet pack mypackage/mypackage.csproj /property:PackageVersion=1.1.5 -o nupkgs -c Release
        nuget push nupkgs/foo.1.1.5.nupkg -src https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json AZ
      displayName: "Pack and push"          
  ```
 

> - In the example below, pushing "MyPackage" version "1.1.3" to an Azure Artifacts feed from an Github action build, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires your pipeline to use the [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication.

Reviewer 2: > In the example below, pushing "MyPackage" version "1.1.3" to an Azure Artifacts feed from an Github action build, the API key "AZ" is only used as a placeholder

This reads a bit funny to me. The subject is missing. 

When writing examples, you are also better off putting a header rather repeating things like `In the example below`.
Try to keep things concise as well.  

An idea would be: 

```
### Pushing from GitHub Actions

When pushing from GitHub actions your pipeline should use the [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication.

---example details---
```

Reviewer: > Try to keep things concise as well.

I agree. Have a look at the examples on other pages. Are they "complete" examples, showing common steps needed to get the point where it can demonstrate the command this page is about, or are they short, showing only the minimum for the current page's concept?

Author: I improved according to your recommendation, please check now.

```cli         
jobs:
  build:
    runs-on: ubuntu-latest

    steps:
    - name: Checkout
      uses: actions/checkout@v3

    - uses: actions/setup-dotnet@v2
      with:
        dotnet-version: 6.0.x # Change to a sdk version matches your solution.
        source-url: https://pkgs.dev.azure.com/<your-organization>/_packaging/<your-feed-name>/nuget/v3/index.json
      env:
        NUGET_AUTH_TOKEN: ${{secrets.AZURE_DEVOPS_PAT}} # Note, create a secret with this name in Settings
    
    - name: Pack and push
      run: |

>           dotnet build mypackage/mypackage.csproj -c Release

Reviewer 2: similar feedback as the previous one.
